File: pbmx_ffi/java/io/rmf/pbmx/ffi/RawXof.java
package io.rmf.pbmx.ffi;

import java.util.Arrays;
import java.util.List;
import com.sun.jna.Structure;
import com.sun.jna.Pointer;

public class RawXof extends Structure {
    public static class ByValue extends RawXof implements Structure.ByValue {}

    public Pointer data;
    public Pointer vtable;

    public ByValue val() {
        ByValue value = new ByValue();
        value.data = this.data;
        value.vtable = this.vtable;
        return value;
    }

    public RawXof ref() {
        RawXof x = new RawXof();
        x.data = this.data;
        x.vtable = this.vtable;
        return x;
    }

    @Override
    protected List getFieldOrder() {
        return Arrays.asList(new String[] { "data", "vtable" });
    }

    @Override
    protected void finalize() {
        LibPbmx.INSTANCE.pbmx_delete_xof(this.val());
    }
}

File: pbmx_droid/app/src/main/java/io/rmf/pbmx/ffi/RawXof.kt
package io.rmf.pbmx.ffi

import com.sun.jna.Pointer
import com.sun.jna.Structure

open class RawXof : Structure() {

    @Suppress("MemberVisibilityCanBePrivate")
    lateinit var data: Pointer
    @Suppress("MemberVisibilityCanBePrivate")
    lateinit var vtable: Pointer

    class ByValue : Structure(), Structure.ByValue {
        lateinit var data: Pointer
        lateinit var vtable: Pointer

        override fun getFieldOrder(): List<String> {
            return listOf("data", "vtable")
        }
    }

    fun value(): ByValue {
        val value = ByValue()
        value.data = this.data
        value.vtable = this.vtable
        return value
    }

    override fun getFieldOrder(): List<String> {
        return listOf("data", "vtable")
    }

    protected fun finalize() {
        FFI.pbmx_delete_xof(this.value())
    }
}

